lib/ListMultiGroupable.js

```diff
@@ -0,0 +1,468 @@
+  }
+}
+
+function _pushNewOrderedKey(obj, key) {
+  var shouldPushKey = obj && obj._orderedKeys && obj._orderedKeys.indexOf(key) === -1;
+
+  shouldPushKey && obj._orderedKeys.push(key);
+}
+
+function _setIn(obj, path, val) {
+  var cloned = babelHelpers._extends({}, obj);
+
+  path.reduce(function (seed, current, idx) {
+    if (idx == path.length - 1) {
+      seed[current] = val;
+    } else if (!seed[current]) {
+      seed[current] = {};
+    }
+
+    _ensureOrderedKeysExists(seed);
+    _pushNewOrderedKey(seed, current);
+
+    return seed[current];
+  }, cloned);
+
+  return cloned;
+}
+
+function _validateOrderedKeyObject(obj) {
+  if (!(obj && obj._orderedKeys)) {
+    throw new Error("currentNode is null/undefined/falsy, or is missing `_orderedKeys`");
+  }
+}
+
+function _pushPathStep(path, nextStep) {
+  if (!path || path.trim() === '') {
+    return nextStep;
+  }
+
+  return _stringifyPath([path, nextStep]);
+}
+
+function _flattenGroups(groups, array) {
+  if (groups && groups._orderedKeys) {
+    groups._orderedKeys.forEach(function (key) {
+      var value = groups[key];
+
+      if (Array.isArray(value)) {
+        value.forEach(function (item) {
+          return array.push(item);
+        });
+      } else {
+        _flattenGroups(value, array);
+      }
+    });
+  }
+}
+
+/*
+state: {
+  depth:     number,
+  offset:    number,
+  traversed: string,
+}
+*/
+function __processHeadersAndItems(currentNode, array, processHeader, processItems, state) {
+  _validateOrderedKeyObject(currentNode);
+
+  return currentNode._orderedKeys.reduce(function (_state, key) {
+    var depth = _state.depth;
+    var offset = _state.offset;
+    var traversed = _state.traversed;
+    var value = currentNode[key];
+
+    var newlyTraversed = _pushPathStep(traversed, key);
+
+    array.push(processHeader(newlyTraversed, key, _state));
+
+    if (!Array.isArray(value)) {
+      var nextState = babelHelpers._extends({}, _state, {
+        depth: depth + 1
+      });
+
+      return __processHeadersAndItems(value, array, processHeader, processItems, nextState);
+    } else {
+      array.push(processItems(value, newlyTraversed, offset, depth + 1));
+
+      return babelHelpers._extends({}, _state, {
+        offset: offset + value.length
+      });
+    }
+  }, state);
+}
+
+function _renderAllTheThings(groupedObj, renderGroupHeader, renderSingleItem) {
+  var outputArray = [];
+  var getChildren = function getChildren(obj) {
+    return obj._orderedKeys;
+  };
+  var onInternal = function onInternal(key, state) {
+    outputArray.push(renderGroupHeader(key, state));
+  };
+  var onLeaf = function onLeaf(array, state) {
+    array.forEach(function (item, idx) {
+      outputArray.push(renderSingleItem(item, state, idx));
+    });
+  };
+
+  _utilObjectTraversal.depthFirst(groupedObj, getChildren, onInternal, onLeaf);
+
+  return outputArray;
+}
+
+function _renderHeadersAndItems(groupedData, processHeader, processItems) {
```

Comment:
Some of these are old functions I forgot to pull out.

Thanks for catching them